Cannot Compile Librespot (Unsure if Specific to Wrapper).

Attempting to run sudo make install for a global install as I intend on using this in my project https://github.com/AmilieCoding/rithmos-rewritten, a C rewrite of my older Rust application.

Unfortunately,  Librespot Rust refuses to compile - I'm unsure as to if this is a Librespot issue or a Librespot-C issue.

```error[E0308]: mismatched types
  --> playback.rs:51:17
   |
50 |               Mixer(Box::into_raw(Box::new(
   |                                   -------- arguments to this function are incorrect
51 | /                 mixer::find(Some(
52 | |                     CStr::from_ptr(mixer_name).to_str().unwrap()
53 | |                 )).expect("Failed to find mixer.")(*Box::from_raw((*mixer_config).0))
   | |_____________________________________________________________________________________^ expected `Arc<dyn Mixer>`, found `Result<Arc<dyn Mixer>, Error>`
   |
   = note: expected struct `Arc<dyn librespot::librespot_playback::mixer::Mixer>`
                found enum `std::result::Result<Arc<dyn librespot::librespot_playback::mixer::Mixer>, librespot::librespot_core::Error>`
note: associated function defined here
  --> /rustc/29483883eed69d5fb4db01964cdf2af4d86e9cb2/library/alloc/src/boxed.rs:260:12
help: consider using `Result::expect` to unwrap the `std::result::Result<Arc<dyn librespot::librespot_playback::mixer::Mixer>, librespot::librespot_core::Error>` value, panicking if the value is a `Result::Err`
   |
53 |                 )).expect("Failed to find mixer.")(*Box::from_raw((*mixer_config).0)).expect("REASON")
   |                                                                                      +++++++++++++++++

error[E0063]: missing field `local_file_directories` in initializer of `librespot::librespot_playback::config::PlayerConfig`
  --> playback.rs:99:13
   |
99 |             config::PlayerConfig {
   |             ^^^^^^^^^^^^^^^^^^^^ missing `local_file_directories`

error[E0599]: no function or associated item named `from_uri` found for struct `librespot::librespot_core::SpotifyId` in the current scope
   --> playback.rs:178:32
    |
178 |         match core::SpotifyId::from_uri(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
    |                                ^^^^^^^^ function or associated item not found in `librespot::librespot_core::SpotifyId`
    |
note: if you're trying to build a new `librespot::librespot_core::SpotifyId` consider using one of the following associated functions:
      librespot::librespot_core::SpotifyId::from_base16
      librespot::librespot_core::SpotifyId::from_base62
      librespot::librespot_core::SpotifyId::from_raw
   --> /root/.cargo/git/checkouts/librespot-0af58e53e65aac12/3eca1ab/core/src/spotify_id.rs:42:5
    |
42  |     pub fn from_base16(src: &str) -> SpotifyIdResult {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
56  |     pub fn from_base62(src: &str) -> SpotifyIdResult {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
80  |     pub fn from_raw(src: &[u8]) -> SpotifyIdResult {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: there is an associated function `from` with a similar name
    |
178 -         match core::SpotifyId::from_uri(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
178 +         match core::SpotifyId::from(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
    |

error[E0599]: no function or associated item named `from_uri` found for struct `librespot::librespot_core::SpotifyId` in the current scope
   --> playback.rs:199:32
    |
199 |         match core::SpotifyId::from_uri(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
    |                                ^^^^^^^^ function or associated item not found in `librespot::librespot_core::SpotifyId`
    |
note: if you're trying to build a new `librespot::librespot_core::SpotifyId` consider using one of the following associated functions:
      librespot::librespot_core::SpotifyId::from_base16
      librespot::librespot_core::SpotifyId::from_base62
      librespot::librespot_core::SpotifyId::from_raw
   --> /root/.cargo/git/checkouts/librespot-0af58e53e65aac12/3eca1ab/core/src/spotify_id.rs:42:5
    |
42  |     pub fn from_base16(src: &str) -> SpotifyIdResult {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
56  |     pub fn from_base62(src: &str) -> SpotifyIdResult {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
80  |     pub fn from_raw(src: &[u8]) -> SpotifyIdResult {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: there is an associated function `from` with a similar name
    |
199 -         match core::SpotifyId::from_uri(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
199 +         match core::SpotifyId::from(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
    |

error[E0308]: mismatched types
   --> playback.rs:287:63
    |
287 | ...                   track_id: spotify_id_new_internal(track_id)
    |                                 ----------------------- ^^^^^^^^ expected `SpotifyId`, found `SpotifyUri`
    |                                 |
    |                                 arguments to this function are incorrect
    |
note: function defined here
   --> core.rs:83:8
    |
83  | pub fn spotify_id_new_internal(spotify_id: core::SpotifyId) -> *mut SpotifyId {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ ---------------------------

error[E0308]: mismatched types
   --> playback.rs:300:63
    |
300 | ...                   track_id: spotify_id_new_internal(track_id),
    |                                 ----------------------- ^^^^^^^^ expected `SpotifyId`, found `SpotifyUri`
    |                                 |
    |                                 arguments to this function are incorrect
    |
note: function defined here
   --> core.rs:83:8
    |
83  | pub fn spotify_id_new_internal(spotify_id: core::SpotifyId) -> *mut SpotifyId {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ ---------------------------

error[E0308]: mismatched types
   --> playback.rs:311:63
    |
311 | ...                   track_id: spotify_id_new_internal(track_id)
    |                                 ----------------------- ^^^^^^^^ expected `SpotifyId`, found `SpotifyUri`
    |                                 |
    |                                 arguments to this function are incorrect
    |
note: function defined here
   --> core.rs:83:8
    |
83  | pub fn spotify_id_new_internal(spotify_id: core::SpotifyId) -> *mut SpotifyId {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ ---------------------------

error[E0308]: mismatched types
   --> playback.rs:324:63
    |
324 | ...                   track_id: spotify_id_new_internal(track_id),
    |                                 ----------------------- ^^^^^^^^ expected `SpotifyId`, found `SpotifyUri`
    |                                 |
    |                                 arguments to this function are incorrect
    |
note: function defined here
   --> core.rs:83:8
    |
83  | pub fn spotify_id_new_internal(spotify_id: core::SpotifyId) -> *mut SpotifyId {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ ---------------------------

error[E0308]: mismatched types
   --> playback.rs:338:63
    |
338 | ...                   track_id: spotify_id_new_internal(track_id),
    |                                 ----------------------- ^^^^^^^^ expected `SpotifyId`, found `SpotifyUri`
    |                                 |
    |                                 arguments to this function are incorrect
    |
note: function defined here
   --> core.rs:83:8
    |
83  | pub fn spotify_id_new_internal(spotify_id: core::SpotifyId) -> *mut SpotifyId {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ ---------------------------

error[E0308]: mismatched types
   --> playback.rs:351:63
    |
351 | ...                   track_id: spotify_id_new_internal(track_id)
    |                                 ----------------------- ^^^^^^^^ expected `SpotifyId`, found `SpotifyUri`
    |                                 |
    |                                 arguments to this function are incorrect
    |
note: function defined here
   --> core.rs:83:8
    |
83  | pub fn spotify_id_new_internal(spotify_id: core::SpotifyId) -> *mut SpotifyId {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ ---------------------------

error[E0308]: mismatched types
   --> playback.rs:363:63
    |
363 | ...                   track_id: spotify_id_new_internal(track_id)
    |                                 ----------------------- ^^^^^^^^ expected `SpotifyId`, found `SpotifyUri`
    |                                 |
    |                                 arguments to this function are incorrect
    |
note: function defined here
   --> core.rs:83:8
    |
83  | pub fn spotify_id_new_internal(spotify_id: core::SpotifyId) -> *mut SpotifyId {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ ---------------------------

error[E0308]: mismatched types
   --> playback.rs:376:63
    |
376 | ...                   track_id: spotify_id_new_internal(track_id),
    |                                 ----------------------- ^^^^^^^^ expected `SpotifyId`, found `SpotifyUri`
    |                                 |
    |                                 arguments to this function are incorrect
    |
note: function defined here
   --> core.rs:83:8
    |
83  | pub fn spotify_id_new_internal(spotify_id: core::SpotifyId) -> *mut SpotifyId {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ ---------------------------

Some errors have detailed explanations: E0063, E0308, E0599.
For more information about an error, try `rustc --explain E0063`.
error: could not compile `librespot` (lib) due to 12 previous errors
make[2]: *** [CMakeFiles/librespot_rust.dir/build.make:70: CMakeFiles/librespot_rust] Error 101
make[1]: *** [CMakeFiles/Makefile2:120: CMakeFiles/librespot_rust.dir/all] Error 2
make: *** [Makefile:136: all] Error 2```

--- a/core/core.rs
+++ b/core/core.rs
@@ -78,25 +78,25 @@ pub fn session_connect(session: *mut Session, credentials: *mut Credentials) {
 }
 
 #[repr(C)]
-pub struct SpotifyId(*mut core::SpotifyId);
+pub struct SpotifyUri(*mut core::SpotifyUri);
 
-pub fn spotify_id_new_internal(spotify_id: core::SpotifyId) -> *mut SpotifyId {
+pub fn spotify_uri_new_internal(spotify_uri: core::SpotifyUri) -> *mut SpotifyUri {
     Box::into_raw(Box::new(
-        SpotifyId(Box::into_raw(Box::new(
-            spotify_id
+        SpotifyUri(Box::into_raw(Box::new(
+            spotify_uri
         )))
     ))
 }
 
 #[no_mangle]
-pub fn spotify_id_new() {}
+pub fn spotify_uri_new() {}
 
 #[no_mangle]
-pub fn spotify_id_free(spotify_id: *mut SpotifyId) {
-    if spotify_id.is_null() {
+pub fn spotify_uri_free(spotify_uri: *mut SpotifyUri) {
+    if spotify_uri.is_null() {
         return;
     }
     unsafe {
-        drop(Box::from_raw(spotify_id));
+        drop(Box::from_raw(spotify_uri));
     }
 }

--- a/core/oauth.rs
+++ b/core/oauth.rs
@@ -129,7 +129,7 @@ pub fn oauth_build(oauth_builder: *mut OAuthBuilder) -> *mut OAuth {
             CStr::from_ptr((*oauth_builder).redirect_uri).to_str().unwrap(),
             (*oauth_builder).scopes.iter().map(|s| s.as_str()).collect()
         );
-        if !(*oauth_builder).message.is_empty() {
+        if !(&(*oauth_builder)).message.is_empty() {
             librespot_oauth_builder = librespot_oauth_builder.with_custom_message(&(*oauth_builder).message);
         }
         if (*oauth_builder).auto_open {

--- a/core/playback.rs
+++ b/core/playback.rs
@@ -12,7 +12,7 @@ use std::ffi::{c_char, c_uchar, CStr};
 use std::sync::Arc;
 use std::time::Duration;
 
-use crate::core::{session_box, spotify_id_new_internal, Session, SpotifyId};
+use crate::core::{session_box, spotify_uri_new_internal, Session, SpotifyUri};
 use crate::runtime::runtime;
 
 #[repr(C)]
@@ -50,7 +50,7 @@ pub fn mixer_new(mixer_config: *mut MixerConfig, mixer_name: *const c_char) -> *
             Mixer(Box::into_raw(Box::new(
                 mixer::find(Some(
                     CStr::from_ptr(mixer_name).to_str().unwrap()
-                )).expect("Failed to find mixer.")(*Box::from_raw((*mixer_config).0))
+                )).expect("Failed to find mixer.")(*Box::from_raw((*mixer_config).0)).expect("Failed to create mixer.")
             )))
         ));
     }
@@ -107,6 +107,7 @@ pub fn player_config_default() -> *mut PlayerConfig {
                 normalisation_attack_cf: 0.005,
                 normalisation_release_cf: 0.1,
                 normalisation_knee_db: 5.0,
+                local_file_directories: vec![],
                 passthrough: false,
                 ditherer: Some(config::mk_ditherer::<config::TriangularDitherer>),
                 position_update_interval: Some(Duration::from_millis(500))
@@ -175,7 +176,7 @@ pub fn player_is_valid(player: *mut Player) -> u8 {
 #[no_mangle]
 pub fn player_load(player: *mut Player, spotify_uri: *const c_char, start_playing: c_uchar, position_ms: u32) {
     unsafe {
-        match core::SpotifyId::from_uri(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
+        match core::SpotifyUri::from_uri(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
             Ok(id) => {
                 if !id.is_playable() {
                     eprintln!("Track is not playable.");
@@ -196,7 +197,7 @@ pub fn player_load(player: *mut Player, spotify_uri: *const c_char, start_playin
 #[no_mangle]
 pub fn player_preload(player: *mut Player, spotify_uri: *const c_char) {
     unsafe {
-        match core::SpotifyId::from_uri(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
+        match core::SpotifyUri::from_uri(CStr::from_ptr(spotify_uri).to_str().unwrap()) {
             Ok(id) => {
                 if !id.is_playable() {
                     eprintln!("Track is not playable.");
@@ -284,7 +285,7 @@ pub fn player_channel_poll(player_channel: *mut PlayerChannel, player_event: *mu
                         (*player_event).event = PlayerEventType::PLAYER_EVENT_STOPPED;
                         (*player_event).data.stopped = PlayerEventStopped {
                             play_request_id,
-                            track_id: spotify_id_new_internal(track_id)
+                            track_id: spotify_uri_new_internal(track_id)
                         };
                         return true as u8;
                     }
@@ -297,7 +298,7 @@ pub fn player_channel_poll(player_channel: *mut PlayerChannel, player_event: *mu
                         (*player_event).event = PlayerEventType::PLAYER_EVENT_LOADING;
                         (*player_event).data.loading = PlayerEventLoading {
                             play_request_id,
-                            track_id: spotify_id_new_internal(track_id),
+                            track_id: spotify_uri_new_internal(track_id),
                             position_ms
                         };
                         return true as u8;
@@ -308,7 +309,7 @@ pub fn player_channel_poll(player_channel: *mut PlayerChannel, player_event: *mu
                     } => {
                         (*player_event).event = PlayerEventType::PLAYER_EVENT_PRELOADING;
                         (*player_event).data.preloading = PlayerEventPreloading {
-                            track_id: spotify_id_new_internal(track_id)
+                            track_id: spotify_uri_new_internal(track_id)
                         };
                         return true as u8;
                     }
@@ -321,7 +322,7 @@ pub fn player_channel_poll(player_channel: *mut PlayerChannel, player_event: *mu
                         (*player_event).event = PlayerEventType::PLAYER_EVENT_PLAYING;
                         (*player_event).data.playing = PlayerEventPlaying {
                             play_request_id,
-                            track_id: spotify_id_new_internal(track_id),
+                            track_id: spotify_uri_new_internal(track_id),
                             position_ms
                         };
                         return true as u8;
@@ -335,7 +336,7 @@ pub fn player_channel_poll(player_channel: *mut PlayerChannel, player_event: *mu
                         (*player_event).event = PlayerEventType::PLAYER_EVENT_PAUSED;
                         (*player_event).data.paused = PlayerEventPaused {
                             play_request_id,
-                            track_id: spotify_id_new_internal(track_id),
+                            track_id: spotify_uri_new_internal(track_id),
                             position_ms
                         };
                         return true as u8;
@@ -348,7 +349,7 @@ pub fn player_channel_poll(player_channel: *mut PlayerChannel, player_event: *mu
                         (*player_event).event = PlayerEventType::PLAYER_EVENT_END_OF_TRACK;
                         (*player_event).data.end_of_track = PlayerEventEndOfTrack {
                             play_request_id, 
-                            track_id: spotify_id_new_internal(track_id)
+                            track_id: spotify_uri_new_internal(track_id)
                         };
                         return true as u8;
                     }
@@ -360,7 +361,7 @@ pub fn player_channel_poll(player_channel: *mut PlayerChannel, player_event: *mu
                         (*player_event).event = PlayerEventType::PLAYER_EVENT_TIME_TO_PRELOAD_NEXT_TRACK;
                         (*player_event).data.time_to_preload_next_track = PlayerEventTimeToPreloadNextTrack {
                             play_request_id, 
-                            track_id: spotify_id_new_internal(track_id)
+                            track_id: spotify_uri_new_internal(track_id)
                         };
                         return true as u8;
                     }
@@ -373,7 +374,7 @@ pub fn player_channel_poll(player_channel: *mut PlayerChannel, player_event: *mu
                         (*player_event).event = PlayerEventType::PLAYER_EVENT_POSITION_CHANGED;
                         (*player_event).data.positon_changed = PlayerEventPositionChanged {
                             play_request_id,
-                            track_id: spotify_id_new_internal(track_id),
+                            track_id: spotify_uri_new_internal(track_id),
                             position_ms
                         };
                         return true as u8;
@@ -403,28 +404,28 @@ pub struct PlayerEventPlayRequestIdChanged {
 #[derive(Clone, Copy)]
 pub struct PlayerEventStopped {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId
+    pub track_id: *mut SpotifyUri
 }
 
 #[repr(C)]
 #[derive(Clone, Copy)]
 pub struct PlayerEventLoading {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId,
+    pub track_id: *mut SpotifyUri,
     pub position_ms: u32
 }
 
 #[repr(C)]
 #[derive(Clone, Copy)]
 pub struct PlayerEventPreloading {
-    pub track_id: *mut SpotifyId
+    pub track_id: *mut SpotifyUri
 }
 
 #[repr(C)]
 #[derive(Clone, Copy)]
 pub struct PlayerEventPlaying {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId,
+    pub track_id: *mut SpotifyUri,
     pub position_ms: u32
 }
 
@@ -432,7 +433,7 @@ pub struct PlayerEventPlaying {
 #[derive(Clone, Copy)]
 pub struct PlayerEventPaused {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId,
+    pub track_id: *mut SpotifyUri,
     pub position_ms: u32
 }
 
@@ -440,21 +441,21 @@ pub struct PlayerEventPaused {
 #[derive(Clone, Copy)]
 pub struct PlayerEventTimeToPreloadNextTrack {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId
+    pub track_id: *mut SpotifyUri
 }
 
 #[repr(C)]
 #[derive(Clone, Copy)]
 pub struct PlayerEventEndOfTrack {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId
+    pub track_id: *mut SpotifyUri
 }
 
 #[repr(C)]
 #[derive(Clone, Copy)]
 pub struct PlayerEventUnavailable {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId
+    pub track_id: *mut SpotifyUri
 }
 
 #[repr(C)]
@@ -467,7 +468,7 @@ pub struct PlayerEventVolumeChanged {
 #[derive(Clone, Copy)]
 pub struct PlayerEventPositionCorrection {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId,
+    pub track_id: *mut SpotifyUri,
     pub position_ms: u32
 }
 
@@ -475,7 +476,7 @@ pub struct PlayerEventPositionCorrection {
 #[derive(Clone, Copy)]
 pub struct PlayerEventPositionChanged {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId,
+    pub track_id: *mut SpotifyUri,
     pub position_ms: u32
 }
 
@@ -483,7 +484,7 @@ pub struct PlayerEventPositionChanged {
 #[derive(Clone, Copy)]
 pub struct PlayerEventSeeked {
     pub play_request_id: u64,
-    pub track_id: *mut SpotifyId,
+    pub track_id: *mut SpotifyUri,
     pub position_ms: u32
 }
 
@@ -505,7 +506,7 @@ pub union PlayerEventData {
 }
 
 #[repr(C)]
-#[allow(non_camel_case_types)]
+#[allow(non_camel_case_types, dead_code)]
 pub enum PlayerEventType {
     PLAYER_EVENT_NONE,
     PLAYER_EVENT_PLAYER_REQUEST_ID_CHANGED,

--- a/include/librespot/core.h
+++ b/include/librespot/core.h
@@ -16,8 +16,8 @@ extern Session* session_new(SessionConfig* session_config);
 extern void session_free(Session* session);
 extern void session_connect(Session* session, Credentials* credentials);
 
-typedef struct SpotifyId SpotifyId;
-extern void spotify_id_free(SpotifyId* spotify_id);
+typedef struct SpotifyUri SpotifyUri;
+extern void spotify_uri_free(SpotifyUri* spotify_uri);
 
 #ifdef __cplusplus
 }

--- a/include/librespot/playback.h
+++ b/include/librespot/playback.h
@@ -62,44 +62,44 @@ typedef struct {
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
 } PlayerEventStopped;
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
     uint32_t position_ms;
 } PlayerEventLoading;
 
 typedef struct {
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
 } PlayerEventPreloading;
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
     uint32_t position_ms;
 } PlayerEventPlaying;
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
     uint32_t position_ms;
 } PlayerEventPaused;
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
 } PlayerEventTimeToPreloadNextTrack;
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
 } PlayerEventEndOfTrack;
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
 } PlayerEventUnavailable;
 
 typedef struct {
@@ -108,19 +108,19 @@ typedef struct {
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
     uint32_t position_ms;
 } PlayerEventPositionCorrection;
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
     uint32_t position_ms;
 } PlayerEventPositionChanged;
 
 typedef struct {
     uint64_t play_request_id;
-    SpotifyId* track_id;
+    SpotifyUri* track_id;
     uint32_t position_ms;
 } PlayerEventSeeked;
 

